### PR TITLE
doc: add cluster status design

### PR DESCRIPTION
**Pull request by xiang90**
/cc @hongchaodeng

### CODE REVIEW

File: doc/design/cluster_status.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+## Cluster status reporting
+
+An etcd cluster can permanently fail when all members are dead due to massive nodes failure. A user might update etcd cluster TPR with invalid input or format. The operator needs to notify users these bad events.
```
Comment (hongchaodeng) [inline doc/design/cluster_status.md]: "when all members are dead due to massive nodes failure." ->
"when all members are dead and no backup is available."

Comment (hongchaodeng) [inline doc/design/cluster_status.md]: Can we put the last sentence to first:

"The operator needs to notify users bad events. For example, ..."

File: doc/design/cluster_status.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	Status               ClusterStatus `json:"status"`   
+}
+
+Type ClusterStatus struct {
+	Paused bool `json:"paused"`
+	...
+}
+```
+
+The etcd operator keeps the truth of the cluster status, and update the TRP item after each cluster reconciliation atomically. Note that users can modify TPR item concurrently with the operator. The operator MUST NOT overwrite the user input. The user MIGHT update TPR with empty status. The operator MUST handle these two cases gracefully.
```
Comment (hongchaodeng) [inline doc/design/cluster_status.md]: TRP -> TPR

Comment (hongchaodeng) [inline doc/design/cluster_status.md]: How about:
"Note that we have two corner cases: 1. ...; 2. ..."

File: doc/design/cluster_status.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+To not overwrite the user input, the etcd operator will set resource version, and do a compare and swap to update the status. If the compare and swap fails, the operator knows that the user updated the spec concurrently, and it will retry later after get the new spec.
```
Comment (hongchaodeng) [inline doc/design/cluster_status.md]: "To avoid overwriting..."

Comment (hongchaodeng) [inline doc/design/cluster_status.md]: I don't think k8s client supports CAS.

File: doc/design/cluster_status.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+To not be affected by the potential empty or broken status accidently wrote by the user, the etcd operator will never read the status from TPR. It always keeps the source of truth in memory.
```
Comment (hongchaodeng) [inline doc/design/cluster_status.md]: "accidentally written ..."

File: doc/design/cluster_status.md (hunk 3 above)
Comment (xiang90) [inline doc/design/cluster_status.md]: It supports CAS through resource version.

**Comment by xiang90** [comment]
@hongchaodeng All fixed. PTAL.

**Comment by hongchaodeng** [comment]
Just for sync-up, can you clarify how to do CAS? I couldn't find any option in:
https://github.com/kubernetes/kubernetes/blob/850729bfaf192a491c38b9bc15744775e4a0f3b1/pkg/client/clientset_generated/release_1_5/typed/core/v1/pod.go#L34-L42

**Comment by xiang90** [comment]
@hongchaodeng 

For any k8s object, it contains a struct called `ObjectMeta`. Inside this struct, there is a field called `ResourceVersion`. This is used for CAS operations. See the comments:

```
// An opaque value that represents the version of this resource. May be used for optimistic
// concurrency, change detection, and the watch operation on a resource or set of resources.
// Clients must treat these values as opaque and values may only be valid for a particular
// resource or set of resources. Only servers will generate resource versions.
```

We have tested it. It works.

**Comment by hongchaodeng** [comment]
LGTM

**Pull request merged**